// synth-audio/engine/SimpleMusicGenerator.js
import SimpleAudioGenerator from './SimpleAudioGenerator.js';

/**
 * SimpleMusicGenerator - Direct audio synthesis without Web Audio API
 * Generates audio buffers that can be exported to WAV
 */
class SimpleMusicGenerator {
  constructor() {
    this.config = {};
    this.sampleRate = 44100;
  }

  /**
   * Get chord progression based on mood and key
   * Can be overridden with custom chords from AI
   * @returns {Array} Array of chord notes
   */
  getChordProgression() {
    // Custom chords are required
    if (!this.config.customChords) {
      throw new Error('customChords is required in input configuration');
    }
    return this.config.customChords;
  }

  /**
   * Get scale notes based on key
   * Can be overridden with custom scale from AI
   * @returns {Array} Scale notes
   */
  getScale() {
    // Custom scale is required
    if (!this.config.customScale) {
      throw new Error('customScale is required in input configuration');
    }
    return this.config.customScale;
  }

  /**
   * Generate a melody pattern with support for custom melodies
   * @param {number} length - Number of notes to generate
   * @returns {Array} Melody notes
   */
  generateMelody(length = 16) {
    const melodyConfig = this.config.melody;
    
    // Melody is required
    if (!melodyConfig) {
      throw new Error('melody is required in input configuration');
    }
    
    // If melody type is 'custom', use the custom notes
    if (melodyConfig.type === 'custom' && melodyConfig.notes) {
      return melodyConfig.notes;
    }

    throw new Error(`Melody type "${melodyConfig.type}" is not supported in strict mode. Only "custom" melodies with explicit notes are allowed.`);
  }

  /**
   * Generate background music using simple synthesis with advanced control
   * @param {Object} moodAnalysis - Music parameters (mood, key, tempo, etc.)
   * @param {number} duration - Duration in seconds
   * @returns {Float32Array} Audio buffer
   */
  generateBackgroundMusic(moodAnalysis, duration) {
    // All parameters must come from input
    if (!moodAnalysis) throw new Error('moodAnalysis parameters are required');
    if (!duration) throw new Error('duration is required');
    
    this.config = {
      mood: moodAnalysis.mood,
      key: moodAnalysis.key,
      tempo: moodAnalysis.tempo,
      customChords: moodAnalysis.customChords,
      customScale: moodAnalysis.customScale,
      melody: moodAnalysis.melody,
      instruments: moodAnalysis.instruments,
      effects: moodAnalysis.effects,
      dynamics: moodAnalysis.dynamics,
      rests: moodAnalysis.rests || []
    };

    const tempo = this.config.tempo;
    const beatDuration = 60 / tempo;

    // Get sequences - all from input
    const chords = this.getChordProgression();
    const melody = this.generateMelody(Math.ceil(duration / beatDuration));
    const instruments = this.config.instruments; // Required from input
    const rests = this.config.rests || [];

    console.log(`🎼 Generating ${duration}s of music at ${tempo} BPM`);
    console.log(`🎵 Mood: ${moodAnalysis.mood}`);
    console.log(`🔑 Key: ${moodAnalysis.key}`);
    if (this.config.customChords) console.log(`🎼 Using custom chord progression`);
    if (this.config.customScale) console.log(`📊 Using custom scale`);
    if (this.config.melody) console.log(`🎵 Melody type: ${this.config.melody.type || 'random'}`);
    if (rests.length > 0) console.log(`⏸️  Rests configured: ${rests.length}`);

    const audioSamples = [];
    let noteIndex = 0;
    let chordIndex = 0;

    // Helper function to check if we should skip (rest) at this point
    const shouldRest = (currentNoteIndex, currentTime) => {
      for (const rest of rests) {
        if (rest.noteIndex !== null && rest.noteIndex === currentNoteIndex) {
          return rest;
        }
        if (rest.time !== null && Math.abs(currentTime - rest.time) < 0.01) {
          return rest;
        }
      }
      return null;
    };

    // Generate music tracks based on instruments
    for (let time = 0; time < duration; time += beatDuration) {
      // Check if we should insert a rest at this position
      const restConfig = shouldRest(noteIndex, time);
      
      if (restConfig) {
        // Add silence for rest duration
        const restSamples = Math.floor(restConfig.duration * this.sampleRate);
        const silence = new Float32Array(restSamples);
        const positioned = this.positionAudio(silence, time);
        audioSamples.push(positioned);
        console.log(`  └─ ⏸️  ${restConfig.description} at note ${noteIndex} (${restConfig.duration}s)`);
        
        // Skip ahead by rest duration
        noteIndex++;
        chordIndex++;
        continue;
      }

      // Distribute melody and chords across all instruments
      for (let instIdx = 0; instIdx < instruments.length; instIdx++) {
        const instrument = instruments[instIdx];
        
        // Alternate between melody and chords across instruments
        // First half of instruments play melody, second half play chords
        const isMelodyInstrument = instIdx < Math.ceil(instruments.length / 2);
        
        if (isMelodyInstrument) {
          // Generate melody note
          const melodyLoop = this.config.melody && this.config.melody.loop;
          if (melodyLoop || noteIndex < melody.length) {
            const note = melody[noteIndex % melody.length];
            const freq = SimpleAudioGenerator.noteToFrequency(note);
            const waveform = instrument.waveform;
            const volume = instrument.volume;
            
            let noteSamples;
            if (waveform === 'sine') {
              noteSamples = SimpleAudioGenerator.generateSineWave(
                freq, beatDuration * 0.9, this.sampleRate, volume
              );
            } else if (waveform === 'square') {
              noteSamples = SimpleAudioGenerator.generateSquareWave(
                freq, beatDuration * 0.9, this.sampleRate, volume
              );
            } else if (waveform === 'sawtooth') {
              noteSamples = SimpleAudioGenerator.generateSawtoothWave(
                freq, beatDuration * 0.9, this.sampleRate, volume
              );
            } else {
              // Default: triangle
              noteSamples = SimpleAudioGenerator.generateTriangleWave(
                freq, beatDuration * 0.9, this.sampleRate, volume
              );
            }
            
            // Apply envelope if defined
            if (instrument.envelope) {
              noteSamples = this.applyEnvelope(noteSamples, instrument.envelope);
            }
            
            const positioned = this.positionAudio(noteSamples, time);
            audioSamples.push(positioned);
          }
        } else {
          // Generate chord/bass note (chords always loop to fill the duration)
          const chord = chords[chordIndex % chords.length];
          const freq = SimpleAudioGenerator.noteToFrequency(chord);
          const waveform = instrument.waveform;
          const volume = instrument.volume;

          let noteSamples;
          if (waveform === 'triangle') {
            noteSamples = SimpleAudioGenerator.generateTriangleWave(
              freq, beatDuration * 1.5, this.sampleRate, volume
            );
          } else if (waveform === 'square') {
            noteSamples = SimpleAudioGenerator.generateSquareWave(
              freq, beatDuration * 1.5, this.sampleRate, volume
            );
          } else if (waveform === 'sawtooth') {
            noteSamples = SimpleAudioGenerator.generateSawtoothWave(
              freq, beatDuration * 1.5, this.sampleRate, volume
            );
          } else {
            // Default: sine
            noteSamples = SimpleAudioGenerator.generateSineWave(
              freq, beatDuration * 1.5, this.sampleRate, volume
            );
          }

          // Apply envelope if defined
          if (instrument.envelope) {
            noteSamples = this.applyEnvelope(noteSamples, instrument.envelope);
          }

          const positioned = this.positionAudio(noteSamples, time);
          audioSamples.push(positioned);
        }
      }

      noteIndex++;
      chordIndex++;
    }

    console.log(`🎙️  Rendering audio buffer...`);

    // Mix all audio
    if (audioSamples.length === 0) {
      return new Float32Array(Math.floor(duration * this.sampleRate));
    }

    const mixedAudio = this.mixAudioTracks(audioSamples, duration);
    console.log(`✅ Audio buffer created (${(mixedAudio.length / this.sampleRate).toFixed(2)}s)`);

    return mixedAudio;
  }

  /**
   * Apply ADSR envelope to audio samples
   * @param {Float32Array} samples - Audio samples
   * @param {Object} envelope - {attack, decay, sustain, release} in seconds
   * @returns {Float32Array} Enveloped audio
   */
  applyEnvelope(samples, envelope) {
    const { attack = 0.01, decay = 0.1, sustain = 0.5, release = 0.5 } = envelope;
    const totalDuration = samples.length / this.sampleRate;
    
    const attackSamples = Math.floor(attack * this.sampleRate);
    const decaySamples = Math.floor(decay * this.sampleRate);
    const releaseSamples = Math.floor(release * this.sampleRate);
    const sustainStart = attackSamples + decaySamples;
    const sustainEnd = samples.length - releaseSamples;

    for (let i = 0; i < samples.length; i++) {
      let envelope_value = 1.0;

      if (i < attackSamples) {
        // Attack phase
        envelope_value = i / attackSamples;
      } else if (i < sustainStart) {
        // Decay phase
        envelope_value = 1.0 - ((i - attackSamples) / decaySamples) * (1.0 - sustain);
      } else if (i < sustainEnd) {
        // Sustain phase
        envelope_value = sustain;
      } else {
        // Release phase
        envelope_value = sustain * (1.0 - (i - sustainEnd) / releaseSamples);
      }

      samples[i] *= envelope_value;
    }

    return samples;
  }

  /**
   * Position audio samples at a specific time
   * @param {Float32Array} samples - Audio samples
   * @param {number} startTime - Start time in seconds
   * @returns {object} {startIndex, samples}
   */
  positionAudio(samples, startTime) {
    return {
      startIndex: Math.floor(startTime * this.sampleRate),
      samples: samples
    };
  }

  /**
   * Mix multiple positioned audio tracks
   * @param {Array} audioSamples - Array of {startIndex, samples}
   * @param {number} duration - Total duration in seconds
   * @returns {Float32Array} Mixed audio
   */
  mixAudioTracks(audioSamples, duration) {
    const totalSamples = Math.floor(duration * this.sampleRate);
    const mixed = new Float32Array(totalSamples);

    for (const { startIndex, samples } of audioSamples) {
      for (let i = 0; i < samples.length && startIndex + i < totalSamples; i++) {
        mixed[startIndex + i] += samples[i];
      }
    }

    // Normalize to prevent clipping
    let max = 0;
    for (let i = 0; i < mixed.length; i++) {
      max = Math.max(max, Math.abs(mixed[i]));
    }

    if (max > 1) {
      for (let i = 0; i < mixed.length; i++) {
        mixed[i] /= max;
      }
    }

    return mixed;
  }
}

export default SimpleMusicGenerator;
